src/libmongoc/src/mongoc/mongoc-bulkwrite.c

```diff
@@ -0,0 +1,1833 @@
+                        // Check for an optional `upserted._id`.
+                        const bson_value_t *upserted_id = NULL;
+                        bson_iter_t id_iter;
+                        if (bson_iter_init_find (&result_iter, result, "upserted")) {
+                           BSON_ASSERT (bson_iter_init (&result_iter, result));
+                           if (!bson_iter_find_descendant (&result_iter, "upserted._id", &id_iter)) {
+                              bson_set_error (&error,
+                                              MONGOC_ERROR_COMMAND,
+                                              MONGOC_ERROR_COMMAND_INVALID_ARG,
+                                              "expected `upserted` to be a document "
+                                              "containing `_id`, but did not find `_id`");
+                              _bulkwriteexception_set_error (ret.exc, &error);
+                              goto batch_fail;
+                           }
+                           upserted_id = bson_iter_value (&id_iter);
+                        }
+
+                        _bulkwriteresult_set_updateresult (ret.res, n, nModified, upserted_id, models_idx);
+                        break;
+                     }
+                     case MODEL_OP_DELETE: {
+                        // Parse `n`.
+                        int32_t n;
+                        if (!lookup_int32 (result, "n", &n, "result", ret.exc)) {
+                           goto batch_fail;
+                        }
+
+                        _bulkwriteresult_set_deleteresult (ret.res, n, models_idx);
+                        break;
+                     }
+                     case MODEL_OP_INSERT: {
+                        _bulkwriteresult_set_insertresult (ret.res, &md->id_iter, models_idx);
+                        break;
+                     }
+                     default:
+                        // Add an unreachable default case to silence `switch-default` warnings.
+                        BSON_UNREACHABLE ("unexpected default");
+                     }
+                  }
+               }
+               // Ensure iterating cursor did not error.
+               {
+                  const bson_t *error_document;
+                  if (mongoc_cursor_error_document (reply_cursor, &error, &error_document)) {
+                     _bulkwriteexception_set_error (ret.exc, &error);
+                     if (error_document) {
+                        _bulkwriteexception_set_error_reply (ret.exc, error_document);
+                     }
+                     goto batch_fail;
+                  }
+               }
+            }
+         }
+
+         // Check if stream is valid. `mongoc_cluster_run_retryable_write` may have invalidated stream (e.g. due to
+         // processing an error). If invalid, select a new stream before processing more batches.
+         if (!mongoc_cluster_stream_valid (&self->client->cluster, parts.assembled.server_stream)) {
```

Comment:
How does this related to the earlier comment when you actually invoked `mongoc_cluster_run_retryable_write`?

> A retry occurred. Save the newly created stream to use for subsequent commands.

Is that only for a happy path where a retry was successful and the stream changed? And if so, would we _not_ expect `mongoc_cluster_stream_valid` to return `false` here?